### PR TITLE
Approach 2: Making the lengths array's values type Integer instead of String using generic types

### LINKED ISSUE #49: Length should be digested/stored/represented as integer not string
The `lengths` attribute is currently represented as a string and should be converted to integer.

This might have an impact on the digest calculation as well which means digest currently stored might be wrong.

### LINKED ISSUE #53: Making some classes of the data model generic - related problems and disucssion
In this issue we'll discuss some of the limitations of making some of the classes **generic** in our service. And maybe we can fix them.

### Scope
In our data model, we didn't use the simple _java HashMap_ to store the seqcol object, because we wanted to have a solid representation of the seqcol object in each level. And since the _HashMap_ can accept any kind of elements (key, value), we won't be able to keep track of the content of the seqcol object, at a given level, throughout its lifecycle. 
And using strict, predefined objects, such as the `JSONExtData` and the `JSONLevelOne` objects, forces the seqcol object in any level to respect the attributes in the spec. Besides the help of the IDE (compiler) that will identify each attribute without having to explicitly call them by their name (eg: without having to do `map.get("md5-sequences")`)
**Note: Even if we used java Map class to store the seqcol data, we won't be able to store values with different types (Strings and Integers), unless we make it generic as well.** 
### Problem
The problem in the model is that it treats the `lengths` array as a list of `Strings` instead of a list of `Integers`, and this doesn't respect the seqcol specification. So we need to change the type of the `lengths` array into a list of `Integers`. And this will affect only the seqcol **level 2** object. 
### Current Design
In our current design, the seqcol **level 2** is modeled as follows:
We have a class `SeqColExtendedDataEntity`:
```java
public class SeqColExtendedDataEntity {

    protected String digest; // The level 0 digest

    private JSONExtData extendedSeqColData;

    private AttributeType attributeType;

    private SeqColEntity.NamingConvention namingConvention;
}
```
And the `JSONExtData` class:
```java
public class JSONExtData implements Serializable {
    private List<String> object;

    public JSONExtData(List<String> object){
        this.object = object;
    }
}
```
### Our Approach
Our goal is to make the `JSONExtData` class **generic** to make it able to store either a list of _Integers_, in the case of a `lengths` array, or a list of _Strings_, in the case of other attributes (`sequences`, `names`, etc..). So it will look something like this:
```java
public class JSONExtData<T implements Serializable {
    private List<T> object;

    public JSONExtData(List<T> object){
        this.object = object;
    }
}
```

**However**, this will directly affect the `SeqColExtendedDataEntity`, since it will have to specify the type of the used `JSONExtData` (string or integer). So this will lead to also making the `SeqColExtendedDataEntity` **generic** as well, like this:
```java
public class SeqColExtendedDataEntity<T> {

    protected String digest; // The level 0 digest

    private JSONExtData<T> extendedSeqColData;

    private AttributeType attributeType;

    private SeqColEntity.NamingConvention namingConvention;
}
```

Note that we'll have to refactor all the classes and methods that use these two classes. We'll have to add the type when instantiating.  
For example: 
```java 
JSONExtData<String> seqColSequencesArray = new JSONExtData<>();
```

### PR DESCRIPTION
**Pull request by waterflow80**
closes #53 
closes #49 
This is the second approach of making the `JSONExtData` **generic**. And it is also actually working.

So the new design is as follows:
```mermaid
   classDiagram 
   class JSONExtData~T~ {
   T object
  getObject() T
  }
class SeqColExtendedDataEntity~T~{
    String digest
    JSONExtData~T~ extendedSeqColData
}
JSONExtData <-- SeqColExtendedDataEntity
   
```
I think this model allows us to to easily add new attributes with any type: `String`, `Integer`, `List<String>`, `List<Integer>`, etc... compared to the previous approach #54 , which limits us to using only `List<T>`, and where we only specify the type of the array's elements.

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 1]
```diff
@@ -446,20 +451,20 @@ public List<String> getCommonElementsDistinct(List<String> seqColAFields, List<S
     /**
      * Return the number of common elements between listA and listB
      * Note: Time complexity for this method is about O(n²)*/
-    public Integer getCommonElementsCount(List<String> listA, List<String> listB) {
-        List<String> listALocal = new ArrayList<>(listA); // we shouldn't be making changes on the actual lists
-        List<String> listBLocal = new ArrayList<>(listB);
+    public Integer getCommonElementsCount(List<?> listA, List<?> listB) {
+        List<?> listALocal = new ArrayList<>(listA); // we shouldn't be making changes on the actual lists
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:455]: Can't we convert these to sets and do a set intersection?

**Review by sundarvenkata-EBI — COMMENTED** [review]
Looks good and I like this better than the other design based on how future proof it is...

File: src/main/java/uk/ac/ebi/eva/evaseqcol/entities/JSONIntegerExtData.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package uk.ac.ebi.eva.evaseqcol.entities;
+
+import lombok.Data;
+
+import uk.ac.ebi.eva.evaseqcol.utils.JSONExtData;
+
+import java.util.List;
+
+@Data
+public class JSONIntegerExtData extends JSONExtData<List<Integer>> {
```
[the remark this answers is not in the document]
Comment (apriltuesday) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/entities/JSONIntegerExtData.java:10]: Actually, are we using these anywhere?  It looks like we're still using `JSONExtData<List<Integer>>` elsewhere in the code...

**Review by apriltuesday — COMMENTED** [review]
Agree with Sundar, I think the difference in complexity between the two approaches is minimal and the future-proofing of this one is better.

Overall I feel like the complexity/maintainability is not bad, though it's hard to say for certain having not worked on the implementation (like you have!). There are a lot of type annotations for sure but I feel like they're in places that make sense (e.g. services that "need to know" that `names` is a list of strings, etc.)

File: src/main/java/uk/ac/ebi/eva/evaseqcol/entities/JSONIntegerExtData.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/entities/JSONIntegerExtData.java:10]: Ohh, that's a good point, I apologize.
Initially I used the `JSONIntegerExtData` and the `JSONStringExtData`, but then I realized that I didn't need them.

However I have to rewrite the `toString()` method for the `JSONExtData` again. 

I shouldn't have put the unnecessary classes in the PR's description.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java (hunk 1 above)
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:455]: Yep, I assume so.
I'll do that.

**Comment by tcezard** [comment]
This PR can be closed

**Pull request closed without merging**